src/modules/python/pbs_hooks/PBS_power.PY

```diff
@@ -0,0 +1,191 @@
+"""
+
+__doc__ = """
+$Id$
+
+This hook will activate and deactivate a power profile contained
+in the 'eoe' value for a job.
+"""
+
+import pbs
+import os
+from pbs.v1._pmi_utils import _get_vnodes, _svr_vnode
+
+
+def init_power(event):
+    # Get a Power structure and do the connect.  Reject on failure.
+    try:
+        confvar = "PBS_PMINAME"
+        name = None
+        if confvar in os.environ:
+            name = os.environ[confvar]
+        power = pbs.Power(name)
+        power.connect()
+    except Exception, e:
+        event.reject(str(e))
+    return power
+
+
+def vnodes_enabled(job):
+    # see if power operations are allowed on all job vnodes
+    for vn in _get_vnodes(job):
+        if not _svr_vnode(vn).power_provisioning:
+            pbs.logjobmsg(job.id,
+                "power functionality is disabled on vnode %s" % vn)
+            return False
+    return True
+
+
+# Accept if event not serviceable.
+this_event = pbs.event()
+if this_event.type not in [pbs.EXECJOB_PROLOGUE, pbs.EXECJOB_EPILOGUE,
+                           pbs.EXECJOB_BEGIN, pbs.EXECJOB_END,
+                           pbs.EXECHOST_STARTUP, pbs.EXECHOST_PERIODIC]:
+    pbs.logmsg(pbs.LOG_WARNING,
+        "Event not serviceable for power provisioning.")
+    this_event.accept()
+
+
+# Set eoe values for my node
+if this_event.type == pbs.EXECHOST_STARTUP:
+    from pbs.v1._pmi_utils import _is_node_provisionable
+
+    # Don't connect if the server or sched is running.
+    if not _is_node_provisionable():
+        pbs.logmsg(pbs.LOG_DEBUG,
+            "Provisioning cannot be enabled on this host")
+        this_event.accept()
+    power = init_power(this_event)
+    profiles = power.query(pbs.Power.QUERY_PROFILE)
+    if profiles is not None:
+        me = pbs.get_local_nodename()
+        this_event.vnode_list[me].resources_available[
+            "eoe"] = power._map_profile_names(profiles)
+    power.disconnect()
+    this_event.accept()
+
+
+# Gather energy usage for all jobs
+if this_event.type == pbs.EXECHOST_PERIODIC:
+    # Check if any jobs are running
+    if len(this_event.job_list) == 0:
+        this_event.accept()
+
+    power = init_power(this_event)
+    for jobid in this_event.job_list.keys():
+        # set energy usage
+        job = this_event.job_list[jobid]
+        if not job.in_ms_mom():		# skip any jobs that MOM is not MS
+            continue
+        if not vnodes_enabled(job):	# skip if vnodes have power_provisioning=0
+            continue
+
+        try:
+            usage = power.get_usage(job)
+            if usage is not None:
+                job.resources_used["energy"] = usage
+        except Exception, e:
+            pbs.logmsg(pbs.LOG_ERROR, str(e))
+    power.disconnect()
+    this_event.accept()
+
+# From this point on, the event will have a job.
+this_job = this_event.job
+
+# The following code is needed until spid 329611 is fixed.
+if this_event.type == pbs.EXECJOB_BEGIN:
+    me = pbs.get_local_nodename()
```

Comment:
I did not find anywhere documented what exactly it returns. I have seen it mostly returning short names regardless of nodes created either with shortname or FQDN.
This is what guide says. 
6.12.14.3.ix Method to Get Local Vnode Name
pbs.get_local_nodename()
This returns a Python str whose value is the name of the local natural vnode.
If you want to refer to the vnode object representing the current host, you can pass this vnode name as the key to pbs.event().vnode_list[]. For example:
Vn = pbs.event().vnode_list[pbs.get_local_nodename()]